src/System/Diagnostic/SourceCodeIntegrityChecker.php

```diff
@@ -0,0 +1,248 @@
+        // Get files altered
+        $altered = [];
+        foreach ($baseline['files'] as $file => $hash) {
+            if (isset($current['files'][$file]) && $current['files'][$file] !== $hash) {
+                $altered[$file] = $current['files'][$file];
+            }
+        }
+        // Summary where the key is the file and the value is the status. Ignore OK files
+        $summary = [];
+        foreach ($altered as $file => $hash) {
+            $summary[$file] = self::STATUS_ALTERED;
+        }
+        foreach ($added as $file => $hash) {
+            $summary[$file] = self::STATUS_ADDED;
+        }
+        foreach ($missing as $file => $hash) {
+            $summary[$file] = self::STATUS_MISSING;
+        }
+
+        return $summary;
+    }
+
+    private function getGLPIRelease(&$errors = []): string|null
+    {
+        $gh_releases_endpoint = 'https://api.github.com/repos/glpi-project/glpi/releases';
+        $version_to_get = VersionParser::getNormalizedVersion(GLPI_VERSION);
+        $client = new Client();
+        $dest = null;
+        try {
+            $response = $client->request('GET', $gh_releases_endpoint);
+            $releases = json_decode((string) $response->getBody(), true, 512, JSON_THROW_ON_ERROR);
+            foreach ($releases as $release) {
+                if ($release['tag_name'] === $version_to_get) {
+                    foreach ($release['assets'] as $asset) {
+                        if (str_starts_with($asset['name'], 'glpi-' . $version_to_get)) {
+                            $dest = GLPI_TMP_DIR . '/' . $asset['name'];
+                            // Check if the file already exists in the tmp dir
+                            if (file_exists($dest)) {
+                                break;
+                            }
+                            $url = $asset['browser_download_url'];
+                            $client->request('GET', $url, ['sink' => $dest]);
+                            break;
+                        }
+                    }
+                    break;
+                }
+            }
+        } catch (GuzzleException $e) {
+            $errors[] = $e->getMessage();
+        }
+        return $dest;
+    }
+
+    public function getDiff(array &$errors = [], bool $allow_download = false): string
+    {
+        $summary = $this->getSummary();
+        // ignore OK files in case they are present
+        $summary = array_filter($summary, static fn ($status) => $status !== self::STATUS_OK);
+        // Ensure the release is downloaded
+        if ($allow_download) {
+            $release_path = $this->getGLPIRelease($errors);
+        } else {
+            $release_path = GLPI_TMP_DIR . '/glpi-' . VersionParser::getNormalizedVersion(GLPI_VERSION) . '.tgz';
+            if (!file_exists($release_path)) {
+                $errors[] = 'The release is not downloaded and downloading it was not allowed';
+                return '';
+            }
+        }
```

Comment:
When using `bin/console diagnostic:source_code_integrity --diff`, the `interact()` method will set `$allow_download ` to `true` is user answers `yes` (this will probably be the case).

On first execution, release will be downloaded to be able to generate the diff.
On next executions, even if the release is already present on disk, it will be downloaded again. I think this part could be optimized.

Reply:
That shouldn't be the case. `getGLPIRelease` is called every time a download is allowed, but that method doesn't always re-download the archive. Instead, it:
1. Makes a single request to the GitHub API to get a list of releases
2. Checks the releases for the name of the archive related to the current version (in case extensions change at some point)
3. Checks if the correct archive is already downloaded. If yes, the path to the downloaded file is returned without re-downloading. If no, it attempts to download it before returning the expected path.